### DISCUSSION TITLE
Conversion failed when converting from a character string to uniqueidentifier.

### DISCUSSION THREAD

**Discussion by logtved** [Budibase/budibase#6386] [Help] [answered]
Hello

When creating a new data provider and try to pass the UUID from the URL to a new dataprovider it seems it cannot convert from string to uniqueidentifier. 
The Data source is MS SQL

Am I doing something wrong or is that not possible?

**Comment by melohagan** [comment]
Hello @logtved 

Is this a data provider filter that is using the `{{ URL.id }}` binding? 

I had Postgres handy, and I had no issues there with the UUID data type, but it does seem to be interoperable with strings: 
<img width="832" alt="Screenshot 2022-06-20 at 10 57 37" src="https://user-images.githubusercontent.com/101575380/174577211-9706da38-6b2c-43bd-b858-aeee061b2d0a.png">

Maybe with MS SQL strings cannot be used without conversion with UUID column types?

**Reply by logtved** [reply]
Hi @melohagan 

Thanks for the reply. 
Yes it is using the `{{ URL.id }}` binding even as a value instead: `{{ [url].[id] }}` it gives the same error. 
So it seems to pass the id along.

However the SQL Table needs to have a primary key constraint and that is set to the UUID and the data type is set to uniqueidentifier.
but I have not found a way to convert the string to uniqueidentifier. I'm not sure our Budibase sends the Query towards the SQL DB.

**Comment by melohagan** [comment]
@logtved You may need to use custom queries (https://docs.budibase.com/docs/custom-queries)

Then you can pass through the `{{ URL.id }}` binding (see linked doc page) and then you can do the convert within then SQL query statement:
https://docs.microsoft.com/en-us/sql/t-sql/data-types/uniqueidentifier-transact-sql?view=sql-server-ver16#examples

**Comment by logtved** [comment]
@melohagan I already tried that. but I cannot get it to work, as either I cannot get it to convert and get the error as described. 
I tried multiple queries. 
![Error](https://user-images.githubusercontent.com/76409588/174583224-0136d2d0-84b6-4a3d-8240-3ab183be2fc7.jpg)
)

**Reply by melohagan** [reply]
Thanks for the screenshot! You need to pass the binding through from the data provider.

So if you look at this section: https://docs.budibase.com/docs/custom-queries#using-query-bindings

1. Click add binding (call it urlid)
2. Use the binding in your query box `{{ urlid }}`
3. In the **design** section, click the cog icon of the data source, and pass `{{ URL.id }}` as the value for the *urlid* binding.

**Reply by logtved** [reply]
@melohagan 
I tried your steps but it seems I still get the error? perhaps I'm during something wrong? 
![urlid](https://user-images.githubusercontent.com/76409588/174587516-1c3d9ac2-f2d1-40f7-a2fd-05616e3d5656.jpg)
![dataprovider](https://user-images.githubusercontent.com/76409588/174587511-e6b8d044-e1d8-4cce-8e05-e66dcf14093e.jpg)

**Comment by logtved** [comment]
and I cannot save when I cannot execute it .
![Error2](https://user-images.githubusercontent.com/76409588/174588578-cb113a9e-259c-46fc-8d09-98e4593daf2e.jpg)

**Reply by melohagan** [reply]
In order to save the query, it must first succeed. Try putting a valid UUID as the default value for *urlid*

I would also make sure that this query definitely works in SQL Server so you can be certain it's a Budibase issue.

**Reply by logtved** [reply]
The Query `select * from vmhosts where uuid = (SELECT CONVERT(uniqueidentifier,'{{urlid}}'))` works fine in the SQL Query from SSMS with the `{{ urlid }}` replaced with a GUID `xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx` returns the value as expected.

I just tested with a headline and it seems the value `{{url.id}}` is returned encapsulated in `['xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx']`
and then the SQL query cannot pass it, as it has the `[' ']`
The below picture is just a headline with text as `{{ URL.id }}`
![text](https://user-images.githubusercontent.com/76409588/174592811-b188fabe-5c13-4893-ba93-96390d680ca5.jpg)

However the error in SSMS is Invalid column name when trying to pass `['xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx']`

Not sure if this assumption is correct?

**Reply by melohagan** [reply]
Ah good find! In that case you may need to do a bit of transformation before passing through the binding. 
So a JavaScript binding like this:
```
return JSON.parse($("URL.id"))[0]
```

**Reply by logtved** [reply]
I'm not sure where to put it? I have added in to my Query under `Transformer` but I cannot see any difference in my result. 

I'm not familiar with JavaScript. so I'm not completely sure on what the query should do. other than selecting character 0 ??

**Reply by melohagan** [reply]
It would go into the binding value within the **designer**

I'm just thinking that isn't necessarily the issue if using a default test value within the query screen didn't work either. 
If you look at the network tab when you click "Run query" are you able to see the SQL that's generated?

**Reply by logtved** [reply]
If I put the JS into the transformer I get `Query Error: $ is not defined` - I have replaced the `{{urlid}}` with a GUID. and if I return the default value `return data` it executes fine.

**Reply by logtved** [reply]
@melohagan any thoughts on my last reply?

**Reply by melohagan** [reply]
Hey @logtved The JS code would need to go into the **designer** section. 

This is what's passed through to the query:
![Screenshot 2022-06-22 at 10 19 32](https://user-images.githubusercontent.com/101575380/174993069-97ba10ef-3927-4aec-9872-d102575721bc.png)
![Screenshot 2022-06-22 at 10 20 21](https://user-images.githubusercontent.com/101575380/174993211-8391586c-d7b3-4fa1-bb61-9b525bdb02a1.png)

With that said, I had mentioned that I would like to see the SQL that's generated when "Run Query" is clicked with a default `{{ urlid }}`
Could help narrow down the issue as well.

**Reply by logtved** [reply]
@melohagan 
ahh did not get that it should be there. 
However when I run the SQL query with `select * from vmhosts where uuid = {{urlid}}` it gives the same error. 
if I replace `{{urlid}}` with a GUID it is working fine. 
See below screenshots. - Let me know if this was not what you meant? 
![Error3](https://user-images.githubusercontent.com/76409588/175004044-1076b820-f6c4-44e6-bb47-681777dd18f3.jpg)
![completed](https://user-images.githubusercontent.com/76409588/175004037-b97e0ee9-bd97-4873-8b27-40099cf39010.jpg)

**Reply by melohagan** [reply]
So for the first screenshot, take the guid string and paste it into the field beside `urlid` (where the placeholder says *Default*)
Then run query to verify that at least works.

**Reply by logtved** [reply]
@melohagan 
Yes that works. but I don't think the issue itself is due to the `{{urlid}}` the issue is that the `{{URL.id}}` has `['']` wrapped around it when it send it as the `{{urlid}}` so that the `GUID` becomes `['GUID']` and SQL cannot parse the `['']`
![completed2](https://user-images.githubusercontent.com/76409588/175006137-2b9d1837-da68-4f0d-90bf-6be28fd597f2.jpg)

**Reply by melohagan** [reply]
Okay good. This is why I was hoping that the following JS would work `return JSON.parse($("URL.id"))[0]`
Because it should grab contents of `['GUID']`
It may also need decoded. If that JS is not working, then could you send me a screenshot of the browser URL when running your app and you're on the "edit" screen.

**Reply by logtved** [reply]
@melohagan 
The JS is just giving me the following `Error while executing JS`
I'm not sure about what URL you mean.
This is when I'm in the design and selected the part where the info should be under the data provider
![url](https://user-images.githubusercontent.com/76409588/175008132-11673187-4bc7-4c1b-a311-547e510cd136.jpg)

and this is from running the app in preview where the data should be presented
![url2](https://user-images.githubusercontent.com/76409588/175008254-df813d6b-4d97-4375-b900-c138e8d0ca8b.jpg)

**Reply by melohagan** [reply]
That's the URL I was looking for, thank you.

It does need to be decoded. Try this JavaScript: `return JSON.parse(decodeURI($("URL.id")))[0]`

**Reply by logtved** [reply]
I get the same error. 
and we agree that the JS should be under the data provider - edit data and for the urlid set the JS function. ?

**Reply by melohagan** [reply]
Yep, under the data provider, click the cog icon, then add the JS code to the binding.

**Reply by logtved** [reply]
Great, that is what I did,  and it gives the same error. :(

**Reply by melohagan** [reply]
What does the route look like for that screen? E.g.

![Screenshot 2022-06-22 at 11 48 10](https://user-images.githubusercontent.com/101575380/175011286-6e3b64bb-05a5-4bfb-97c3-251963daad86.png)

**Reply by logtved** [reply]
/vmhosts/:id

**Reply by melohagan** [reply]
Would you mind sharing your app export to support@budibase.com 
I can take a closer look.

**Reply by logtved** [reply]
I have sent it now. Subject: Github #6386

**Comment by melohagan** [comment, marked as answer]
Thanks @logtved 

Looking at the app you sent me, there seems to still be some confusion about the binding. Make sure you have the following:

![Screenshot 2022-06-22 at 12 05 40](https://user-images.githubusercontent.com/101575380/175014281-19fd41d0-cec3-4eb1-afd4-d3e26784c8ec.png)
**You may want to remove the CONVERT part if it's not needed**

![Screenshot 2022-06-22 at 12 06 08](https://user-images.githubusercontent.com/101575380/175014360-1acf3479-bf11-44df-bfde-e4d1d78b5065.png)
**Click the cog icon here**

![Screenshot 2022-06-22 at 12 07 14](https://user-images.githubusercontent.com/101575380/175014558-fe33b81a-a502-4c2a-9774-215ef71f957a.png)

![Screenshot 2022-06-22 at 12 54 09](https://user-images.githubusercontent.com/101575380/175022922-fd96d1fc-51fb-45c3-846c-9962ca3304df.png)

**Reply by logtved** [reply]
It is still the same error I get. 

The `te` query is: `select * from vmhosts where uuid = '{{urlid}}'` with `urlid` as binding
The convert was put into it as it could have been a reason. 

The dataprovider has `return JSON.parse(decodeURI($("URL.id")))[0]` in the JS section.

**Reply by logtved** [reply]
The Heading I have with the javascript in it, just gives me an `Error while executing JS`

**Reply by melohagan** [reply]
Well the heading would say that because in the designer there is no `URL.id` 
Does it work when you run the app?

**Reply by logtved** [reply]
It is the same error in the designer and preview.

**Reply by melohagan** [reply]
Ah there does seem to be an issue with JSON parse. Try using this JavaScript instead: `return decodeURI($("URL.id")).replace(/[\[''\]]/g, "")`

**Reply by logtved** [reply]
That is Working. Thanks a million.

**Reply by melohagan** [reply]
You're welcome, sorry that was a hassle!

**Reply by logtved** [reply]
Can you make a new reply that I can mark as an answer, so all can see it ? instead of the replies, as it does not seem I can make that as an answer.

**Reply by logtved** [reply]
No worries, we got it solved.
We are still testing the software, and is planning to scale up the use of it for a CMDB. 
But thanks alot.

**Reply by melohagan** [reply]
I updated the binding screenshot and marked as answer.
